### DISCUSSION TITLE
Debugging VerneMQ

### DISCUSSION THREAD

**Discussion by mths1** [vernemq/vernemq#2108] [General]
Hello all 
any hints on how to debug VerneMQ? I would like to use IntelliJ (which works great for developing Erlang btw.), but I was unable to get the debugger working. I was able to connect to a running node via the remote node functionality, but breakpoints did not work (nor anything else in the debugger). 

Someone working with Intellij, and can give me a hint? Or recommend some other IDE?

**Comment by ioolkos** [comment]
@mths1 This is actually a great discussion point, thanks for bringing it up. I see this in relation of helping streamline/improve developer experience itself, it's not just about debugging.
I cannot directly consult on IntelliJ. I think it can use the Erlang Language Server project, for instance (not 100% sure how). In any case, I think you should use IntelliJ because it's a great IDE, that is the solution here should certainly not be to recommend another one (I use VSCode and/or Emacs).

And while you can add whatever you need to a local release, we should work on setting up a "dev" profile (release).
Locally, you could even add `wx`, `observer`, `debugger` & more to the release build. 

The profiles I mention are `relx`/`rebar3` stuff. Here's a list of key words that will matter:
`dev_mode`,` include_erts`, `include_src`, `debug_info` and maybe more.
I'll try to look into this.

As a side note, I often just use `recon_tracing` in a Verne shell:
https://ferd.github.io/recon/recon_trace.html
The `calls` function is very useful, including the `return_trace` variant.

**Comment by ioolkos** [comment]
I noticed that `rebar3_cuttlefish` does not handle `relx` configs with multiple releases. Therefore we cannot add a full developer release as a build target. (we can work with profiles but profiles cannot add applications to a release, only modify options).

The link below has some info on the Erlang language server and debugging. In any case, the `debugger` application needs to be added to the release. This can be done locally anytime, of course, and might be the quickest option of all.  https://erlang-ls.github.io/articles/tutorial-debugger/
If the debugger application is added to the release, it can also be started from within VerneMQ with `debugger:start()`. This will start the debugger GUI.
